src/cargo/ops/cargo_new.rs

```diff
@@ -44,40 +57,231 @@ struct CargoNewConfig {
+        Test { proposed_path: format!("src/main.rs"),     handling: H::Bin },
+        Test { proposed_path: format!("main.rs"),         handling: H::Bin },
+        Test { proposed_path: format!("src/{}.rs", name), handling: H::Detect },
+        Test { proposed_path: format!("{}.rs", name),     handling: H::Detect },
+        Test { proposed_path: format!("src/lib.rs"),      handling: H::Lib },
+        Test { proposed_path: format!("lib.rs"),          handling: H::Lib },
+    ];
+    
+    for i in tests {
+        let pp = i.proposed_path;
+        if !paths::file_already_exists(&path.join(&pp)) {
+            continue;
+        }
+        let sfi = match i.handling {
+            H::Bin => {
+                SourceFileInformation { relative_path: pp, 
+                                        target_name: String::from(project_name), 
+                                        bin: true }
+            },
+            H::Lib => {
+                SourceFileInformation { relative_path: pp, 
+                                        target_name: String::from(project_name), 
+                                        bin: false }
+            },
+            H::Detect => {
+                let content = try!(paths::read(&path.join(pp.clone())));
+                let isbin = content.contains("fn main");
+                SourceFileInformation { relative_path: pp, 
+                                        target_name: String::from(project_name), 
+                                        bin: isbin }
+            },
+        };
+        detected_files.push(sfi);
+    }
+    
+    // Check for duplicate lib attempt
+    
+    let mut previous_lib_relpath : Option<&str> = None;
+    let mut binary_target_names: Vec<&str> = vec![];
+        
+    for i in detected_files {
+        if i.bin {
+            binary_target_names.push(&i.target_name);
+        } else {
+            if let Some(plp) = previous_lib_relpath {
+                return Err(human(format!("Cannot have a project with \
+                                         multiple libraries, \
+                                         found both `{}` and `{}`",
+                                        plp, i.relative_path)));
+            }
+            previous_lib_relpath = Some(&i.relative_path);
+        }
+    }
+    
+    // Check for duplicate binary targets
+    
+    {
+        let l1 = binary_target_names.len();
+        binary_target_names.sort();
+        binary_target_names.dedup();
+        let l2 = binary_target_names.len();
+        
+        if l1 != l2 {
+            return Err(human(format!("Cannot initialize a project \
+                                      where multiple binaries share \
+                                      the guessed name. Unfortunately this \
+                                      implementation can't tell you \
+                                      which source files are involved in \
+                                      a collision.")));
```

Comment:
Oh sorry I think I misinterpreted what was going on here.

For the current detection logic I think that this may want to actually generate an error if more than one binary was found. The only conventional way to store multiple binaries in a Cargo project is to have a bunch of files in `src/bin/*.rs`, but if this instead finds `foo.rs`, `src/main.rs`, and `src/foo.rs` it should generate an error saying that this is ambiguous and it's not clear what the "main binary" is here.